modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj

```diff
@@ -1127,6 +1122,32 @@
     [(str \$ aggr-name) {aggr-group aggr-name}]
     extracted-aggr))
 
+(defn- adjust-distinct-aggregations
+  "This function transforms `aggr-expr'` as in [[expand-aggregations]] so identifiers representing array that is
+  a set of _distinct_ values are wrapped in `{$size...}.
+
+  `aggr-expr` is expected to be a clause that is a result of [[extract-aggregations]]. For details see its docstring.
+
+  Distinct values are computed using the `$addToSet` in a `$group` stage. `$size` transforms them to actual count.
+
+  If there is a need for more _pre-post_ transformations (as per results of [[expand-aggregations]]), this function
+  should be generalized."
+  [[aggr-expr mappings]]
+  (let [distinct-keys (filter (fn [[clause]] (= :distinct clause)) (keys mappings))
+        distinct-vals (into #{}
+                            (comp (map #(get mappings %))
+                                  ;; \$ is added to identifiers so eg. `q~count1` becomes `$q~count1`. Those values
+                                  ;; are used match against `aggr-expr` where identifiers have the prefix.
+                                  (map #(str \$ %)))
+                            distinct-keys)
+        aggr-expr' (walk/postwalk (fn [x]
+                                    (if (and (string? x)
+                                             (distinct-vals x))
+                                      {$size x}
+                                      x))
+                                  aggr-expr)]
+    [aggr-expr' mappings]))
```

Comment:
No need for an extra `let` binding here; this can be inlined into the output clause.

Reply:
In context of the following function, `expand-aggregations`, I've deliberately decided to have an explicit var `aggr-expr'`, to signal _on the first sight_ that returned value has a same shape as what's used in the other function. I've also decided to bind the `distinct-keys` var, despite it could have been omitted by moving filter into transducer composition on the next line. I, subjectively, find this form easier on eyes of future reader.

Reply:
But as I'm not the only future reader I did as suggested :)